[clang] Rework `hasBooleanRepresentation`.

michele-scandale opened the pull request:
This is a follow-up of 13aac46332f607a38067b5ddd466071683b8c255. This commit adjusts the implementation of `hasBooleanRepresentation` to have a similar behavior as the one of `hasIntegerRepresentation`. In particular vector of booleans should be handled in `hasBooleanRepresentation`, while `_Atomic(bool)` should not.

michele-scandale commented on clang/include/clang/AST/Type.h:
```diff
@@ -8623,6 +8624,13 @@ inline bool Type::isIntegralOrEnumerationType() const {
 inline bool Type::isBooleanType() const {
   if (const auto *BT = dyn_cast<BuiltinType>(CanonicalType))
     return BT->getKind() == BuiltinType::Bool;
+  if (const EnumType *ET = dyn_cast<EnumType>(CanonicalType)) {
```
This is to match `isIntegerType` that is used by `hasIntegerRepresentation`.
I'm not entirely sure it makes sense to handle enums here (given also the FIXME comment).
I'll try to see what breaks by removing this.

michele-scandale replied:
From a local run of `check-clang` there are no tests failing with this part of the code.

AaronBallman replied:
> I'm not entirely sure it makes sense to handle enums here

I don't believe it makes sense to report true for an enum type. An enumeration type is never a boolean type in C or in C++. An enumeration can have a boolean underlying type, but it's still an enumeration type rather than a boolean type.

andykaylor commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -2088,13 +2085,22 @@ llvm::Value *CodeGenFunction::EmitToMemory(llvm::Value *Value, QualType Ty) {
     Value = Builder.CreateBitCast(Value, StoreTy);
   }
 
+  if (Ty->hasBooleanRepresentation() || Ty->isBitIntType()) {
```
Is `isExtVectorBoolType()` guaranteed to filter out all the cases where a boolean vector type might get us here? It doesn't seem like the previous code had any sort of handling for vectors of bool other than the `isExtVectorBoolType` handling, but this seems like a potential change in behavior.

michele-scandale replied:
I'll double check. I know there have been changes recently to support a different ABI in the HLSL case.

michele-scandale replied:
Only `ExtVectorType` of booleans are supported, the regular `VectorType` disallow boolean element types (see `Sema::BuildVectorType` and `Sema::BuildExtVectorType`).
 In the `if (Ty->isExtVectorBoolType())` above both kind of vector bool ABIs are handled, so there is no issue there.

andykaylor reviewed (approved):
This looks reasonable to me.

michele-scandale commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -2088,13 +2085,22 @@ llvm::Value *CodeGenFunction::EmitToMemory(llvm::Value *Value, QualType Ty) {
 /// Converts a scalar value from its load/store type (as returned
 /// by convertTypeForLoadStore) to its primary IR type (as returned
 /// by ConvertType).
 llvm::Value *CodeGenFunction::EmitFromMemory(llvm::Value *Value, QualType Ty) {
```
In here it is fine to handle the vector of bool types when`Ty->hasBooleanRepresentation()` as only the non packed ABI will reach that code, and it is fine to generate a `Trunc` (just like in `EmitToMemory` a `ZExt` is generated)

AaronBallman commented on clang/include/clang/AST/Type.h:
```diff
@@ -8623,6 +8624,13 @@ inline bool Type::isIntegralOrEnumerationType() const {
+    // Incomplete enum types are not treated as integer types.
+    // FIXME: In C++, enum types are never integer types.
+    return IsEnumDeclComplete(ET->getDecl()) &&
+           !IsEnumDeclScoped(ET->getDecl()) &&
```
Why does the the enum being scoped matter?

michele-scandale replied:
This is just the same code that exist in `isIntegerType`. I am puzzled by this code as well, but I suppose there is a good reason for it.

rjmccall replied:
C specifies that "enumerated types" are integer types, and C++ does not, and `isIntegerType` is trying to honor that as best it can. It's one of those situations where we need to ask what callers actually want, because in most cases it's probably not this language-specific formal property. If `hasIntegerRepresentation` is really a question about the representation of the type, it should be ignoring scoped-ness of enums. But it seems to be used in Sema in ways that should definitely exclude scoped enums, like the semantic analysis of `__builtin_shuffle_vector` and vector math.

`isBooleanType` should probably continue to check only for `bool`.  It seems reasonable for `hasBooleanRepresentation` to look through enums, though, and it should probably ignore the scoped-ness of the enum.

michele-scandale replied:
> C specifies that "enumerated types" are integer types, and C++ does not, and `isIntegerType` is trying to honor that as best it can. It's one of those situations where we need to ask what callers actually want, because in most cases it's probably not this language-specific formal property. If `hasIntegerRepresentation` is really a question about the representation of the type, it should be ignoring scoped-ness of enums. But it seems to be used in Sema in ways that should definitely exclude scoped enums, like the semantic analysis of `__builtin_shuffle_vector` and vector math.

The check on scoped enum comes from:
```
commit 21673c4e7ec08457b53798b9879b7cc9a5909eb8
Author: Douglas Gregor <dgregor@apple.com>
Date:   Thu May 5 16:13:52 2011 +0000

    Scoped enumerations should not be treated as integer types (in the C
    sense). Fixes <rdar://problem/9366066> by eliminating an inconsistency
    between C++ overloading (which handled scoped enumerations correctly)
    and C binary operator type-checking (which didn't).
    
    llvm-svn: 130924
```

> `isBooleanType` should probably continue to check only for `bool`.

Ok.

> It seems reasonable for `hasBooleanRepresentation` to look through enums, though, and it should probably ignore the scoped-ness of the enum.

I can do that, but I'm not super happy about the inconsistency.

rjmccall replied:
Okay. My understanding is that supporting CodeGen and its dual-representation handling of boolean types is the immediate purpose of this method. "Representation" is a good name for such a method because this really is about lower-level representation.

At a high level, scoped enums with an underlying type of `bool` either need to use i8 as both their memory and their scalar lowered type or they need to be doing toMemory/fromMemory operations when moving to/from memory. I think the latter is probably the better situation overall. So I think we really want this method to ignored scoped-ness of enums.

The immediate problem this creates is that we have a bunch of existing methods that are arguably misnamed: they're named something involving "representation", but they're not really being used to ask about low-level representation. Instead, they're vector-inclusive variants of other predicates. I agree this is a bad place to have ended up, because people will naturally expect all the similarly-named methods to behave consistently.

Since the existing methods are arguably misnamed, I wonder if we could reasonably just change them. Since the difference is (AFAIK) purely to make them vector-inclusive, perhaps we should just remove them and have a `getNonVectorType()` method that clients are expected to call before using the existing predicates.

michele-scandale replied:
>Since the existing methods are arguably misnamed, I wonder if we could reasonably just change them. Since the difference is (AFAIK) purely to make them vector-inclusive, perhaps we should just remove them and have a getNonVectorType() method that clients are expected to call before using the existing predicates.

Just to make sure I'm not misunderstanding things: are you suggesting to just drop the "vector of" portion of the `has*Representation` and move that onto clients of these APIs, right?

I'm not sure if that can work in general. E.g.
```
bool Type::hasIntegerRepresentation() const {
  if (const auto *VT = dyn_cast<VectorType>(CanonicalType))
    return VT->getElementType()->isIntegerType();
  if (CanonicalType->isSveVLSBuiltinType()) {
    const auto *VT = cast<BuiltinType>(CanonicalType);
    return VT->getKind() == BuiltinType::SveBool ||
           (VT->getKind() >= BuiltinType::SveInt8 &&
            VT->getKind() <= BuiltinType::SveUint64);
  }
  if (CanonicalType->isRVVVLSBuiltinType()) {
    const auto *VT = cast<BuiltinType>(CanonicalType);
    return (VT->getKind() >= BuiltinType::RvvInt8mf8 &&
            VT->getKind() <= BuiltinType::RvvUint64m8);
  }

  return isIntegerType();
}
```

This is more than just types where `isIntegerType` is `true`, or vector of such types.

rjmccall replied:
I see. I'm not sure I understand why someone added a bunch of new vector types as individual builtin types (from the names, they sure seem to have element types and otherwise fit the general vector model!), but if that's what we have to deal with, then I suppose I'm just arguing that the methods should be renamed to not say "representation".

AaronBallman commented on clang/lib/AST/Type.cpp:
```diff
@@ -2336,16 +2336,9 @@ bool Type::isArithmeticType() const {
-    return AT->getValueType()->hasBooleanRepresentation();
-
-  return false;
+  if (const auto *VT = dyn_cast<VectorType>(CanonicalType))
```
I missed the PR for adding this function; what does it mean for a type to have a boolean representation? I think the header needs some comments explaining this. Does it mean it's an `i1` in LLVM IR?

michele-scandale replied:
I don't think referring to the how the type is lowered in IR make sense for a function in `Type`.

The overall question should be about this family of functions:
```
  /// Determine whether this type has an integer representation
  /// of some sort, e.g., it is an integer type or a vector.
  bool hasIntegerRepresentation() const;

  /// Determine whether this type has an signed integer representation
  /// of some sort, e.g., it is an signed integer type or a vector.
  bool hasSignedIntegerRepresentation() const;

  /// Determine whether this type has an unsigned integer representation
  /// of some sort, e.g., it is an unsigned integer type or a vector.
  bool hasUnsignedIntegerRepresentation() const;

  /// Determine whether this type has a floating-point representation
  /// of some sort, e.g., it is a floating-point type or a vector thereof.
  bool hasFloatingRepresentation() const;
```

In here I'm just trying to have `hasBooleanRepresentation` be consistent with those.

AaronBallman replied:
All of those are different from your changes here though. In each of those cases, the type *is* of the requested representation or a vector of that type. In this PR, enums are reported as having a boolean representation, which is not something anyone will expect.

rjmccall replied:
Don't aim for consistency. Aim for giving the method a clear and useful definition.

AaronBallman replied:
Sure, that's what I was trying to understand. An enumeration whose underlying type is bool and a bool have the same representation as an i1. So does `unsigned _BitInt(1)`, so does that have a boolean representation?

("representation" is what is throwing me off.)

rjmccall replied:
That was meant as a response to Michele, our reviews just raced.

andykaylor replied:
@AaronBallman In case you haven't dug into the background, I move this function here from CGExpr.cpp (where it was a static function) with the goal of sharing the implementation between the LLVM IR codegen and the ClangIR codegen. It's used a couple of different ways there (and the CIR uses will be equivalent), but I think "will this end up as an i1" is a fair summary.

I just moved the function without changing it in any other way. However, that created the potential problem that it might not have been doing what people familiar with the other has*Representation functions would expect, so the question of consistency, or at least congruency, comes into play for people who weren't already using the function.

AaronBallman replied:
Thanks for the explanation @andykaylor!

rjmccall commented on clang/include/clang/AST/Decl.h:
```diff
@@ -5238,6 +5238,14 @@ inline bool IsEnumDeclScoped(EnumDecl *ED) {
+///
+/// We use this function to break a cycle between the inline definitions in
+/// Type.h and Decl.h.
+inline QualType GetEnumDeclIntegerType(EnumDecl *ED) {
```
We've generally been naming new functions with lower-case letters.

rjmccall commented on clang/include/clang/AST/Type.h:
```diff
@@ -2774,7 +2774,7 @@ class alignas(TypeAlignment) Type : public ExtQualsTypeCommonBase {
 
   /// Determine whether this type has a boolean representation
-  /// of some sort.
+  /// of some sort, e.g., it is a boolean type or a vector thereof.
```
```suggestion
  /// of some sort, i.e., it is a boolean type or a vector thereof.
```

michele-scandale commented:
Putting aside for now the naming issue (https://github.com/llvm/llvm-project/pull/136038#discussion_r2049703902), is there anything left for the changes proposed in this PR?

AaronBallman commented:
> Putting aside for now the naming issue ([#136038 (comment)](https://github.com/llvm/llvm-project/pull/136038#discussion_r2049703902)), is there anything left for the changes proposed in this PR?

Based on:

> ... but I think "will this end up as an i1" is a fair summary.

I think this probably should include 1-bit bit-precise integer types in `hasBooleanRepresentation()`, but it may not be critical.

michele-scandale committed:
[clang] Rework `hasBooleanRepresentation`.

This is a follow-up of 13aac46332f607a38067b5ddd466071683b8c255.
This commit adjusts the implementation of `hasBooleanRepresentation` to
somewhat aligned to `hasIntegerRepresentation`.
In particular vector of booleans should be handled in
`hasBooleanRepresentation`, while `_Atomic(bool)` should not.

AaronBallman reviewed (approved):
LGTM!

michele-scandale commented:
@rjmccall Are you ok landing this change, and defer the renaming to a separate change?

rjmccall commented:
That's fine with me, yeah.
